### PR TITLE
random cleanup + enabling clang/LLVM

### PR DESCRIPTION
**Pull request by ConchuOD**
A bunch of random updates, entirely untested so far.
Some bits of cleanup, followed by Nathan's suggestion of how to cut down on the time do the before/after warning checks - hopefully that'll make it viable to run llvm again.

### CODE REVIEW

**Comment by ConchuOD** [comment]
So, this needs a little bit of sequencing to massage my jenkins pipeline, with the CROSS_COMPILE bit. I think it should be okay, but maybe not.

I did a minor rename along the way to align the name of the.

I ran NIPA in my CI with this & it seemed okay - but the processing time for a patch has doubled thanks to enabling clang/LLVM.
It takes, for an arbitrary set of resources, 30 mins to build allmodconfig with clang & 11 minutes for the second build, where the only change was a comment.
The correspoding times were 20 & 12 for gcc. There's something clearly still wrong with how the rebuilding is done.
It's with ccache AND rebuilding in the same directory so you'd hope that it'd really take very little time to do the rebuild, but that's not the case.

File: tests_rv/patch/build_rv64_clang_allmodconfig/build_rv64_clang_allmodconfig.sh [hunk 1]
```diff
@@ -1,23 +1,79 @@
...
+echo "Tree base:"
+git log -1 --pretty='%h ("%s")' HEAD~
+
+echo "Building the tree with the patch"
+
+tuxmake --wrapper ccache --target-arch riscv -e PATH=$PATH --directory . \
+	--environment=KBUILD_BUILD_TIMESTAMP=@1621270510 \
+	--environment=KBUILD_BUILD_USER=tuxmake --environment=KBUILD_BUILD_HOST=tuxmake \
+	-o $tmpdir0 --toolchain llvm -z none --kconfig allmodconfig -K CONFIG_WERROR=n \
+	W=1 CROSS_COMPILE=riscv64-linux- config default \
```
Comment (ConchuOD) [inline tests_rv/patch/build_rv64_clang_allmodconfig/build_rv64_clang_allmodconfig.sh:30]: Building Guo's comment patch today, this took 28 minutes.

File: tests_rv/patch/build_rv64_clang_allmodconfig/build_rv64_clang_allmodconfig.sh [hunk 2]
```diff
@@ -1,23 +1,79 @@
...
+
+git checkout -q HEAD~
+
+echo "Building the tree before the patch"
+
+tuxmake --wrapper ccache --target-arch riscv -e PATH=$PATH --directory . \
+	--environment=KBUILD_BUILD_TIMESTAMP=@1621270510 \
+	--environment=KBUILD_BUILD_USER=tuxmake --environment=KBUILD_BUILD_HOST=tuxmake \
+	-o $tmpdir0 --toolchain llvm -z none --kconfig allmodconfig -K CONFIG_WERROR=n \
+	W=1 CROSS_COMPILE=riscv64-linux- config default \
```
Comment (ConchuOD) [inline tests_rv/patch/build_rv64_clang_allmodconfig/build_rv64_clang_allmodconfig.sh:52]: This took 11 minutes

File: tests_rv/patch/build_rv64_gcc_allmodconfig/build_rv64_gcc_allmodconfig.sh [hunk 3]
```diff
@@ -25,7 +25,7 @@ tuxmake --wrapper ccache --target-arch riscv -e PATH=$PATH --directory . \
 	--environment=KBUILD_BUILD_TIMESTAMP=@1621270510 \
 	--environment=KBUILD_BUILD_USER=tuxmake --environment=KBUILD_BUILD_HOST=tuxmake \
 	-o $tmpdir0 --toolchain gcc -z none --kconfig allmodconfig -K CONFIG_WERROR=n \
-	-K CONFIG_GCC_PLUGINS=n W=1 \
+	-K CONFIG_GCC_PLUGINS=n W=1 CROSS_COMPILE=riscv64-linux- config default \
```
Comment (ConchuOD) [inline tests_rv/patch/build_rv64_gcc_allmodconfig/build_rv64_gcc_allmodconfig.sh]: This one took 20 minutes

File: tests_rv/patch/build_rv64_gcc_allmodconfig/build_rv64_gcc_allmodconfig.sh [hunk 4]
```diff
@@ -46,7 +46,7 @@ tuxmake --wrapper ccache --target-arch riscv -e PATH=$PATH --directory . \
...
```
Comment (ConchuOD) [inline tests_rv/patch/build_rv64_gcc_allmodconfig/build_rv64_gcc_allmodconfig.sh]: and this took 10

**Review by ConchuOD — COMMENTED** [review]
Added the numbers to the relevant places for context & here's the tuxmake output...

gcc:

W: Support for running offline not available (unshare: unshare failed: Operation not permitted)
I: config: PASS in 0:00:01.839426
I: default: PASS in 0:20:47.301790
I: build output in /ci/workspace/OC_ES_linux-test_riscv-patchwork__4/nipa/trees_rv/fixes/build
W: Support for running offline not available (unshare: unshare failed: Operation not permitted)
I: config: PASS in 0:00:01.428265
I: default: PASS in 0:10:05.979234
I: build output in /ci/workspace/OC_ES_linux-test_riscv-patchwork__4/nipa/trees_rv/fixes/build

llvm:
W: Support for running offline not available (unshare: unshare failed: Operation not permitted)
I: config: PASS in 0:00:03.884311
I: default: PASS in 0:28:19.853824
I: build output in /ci/workspace/OC_ES_linux-test_riscv-patchwork__4/nipa/trees_rv/fixes/build
W: Support for running offline not available (unshare: unshare failed: Operation not permitted)
I: config: PASS in 0:00:02.063537
I: default: PASS in 0:11:47.520979
I: build output in /ci/workspace/OC_ES_linux-test_riscv-patchwork__4/nipa/trees_rv/fixes/build

**Comment by ConchuOD** [comment]
> I ran NIPA in my CI with this & it seemed okay - but the processing time for a patch has doubled thanks to enabling clang/LLVM. It takes, for an arbitrary set of resources, 30 mins to build allmodconfig with clang & 11 minutes for the second build, where the only change was a comment. The correspoding times were 20 & 12 for gcc. There's something clearly still wrong with how the rebuilding is done. It's with ccache AND rebuilding in the same directory so you'd hope that it'd really take very little time to do the rebuild, but that's not the case.

As we said last time Bjorn, it's the stage that looks like "CC [M[ drivers/foo/var/baz.mod.o" that takes forever. I am not sure what the stage is called though...

**Comment by bjoto** [comment]
> > I ran NIPA in my CI with this & it seemed okay - but the processing time for a patch has doubled thanks to enabling clang/LLVM. It takes, for an arbitrary set of resources, 30 mins to build allmodconfig with clang & 11 minutes for the second build, where the only change was a comment. The correspoding times were 20 & 12 for gcc. There's something clearly still wrong with how the rebuilding is done. It's with ccache AND rebuilding in the same directory so you'd hope that it'd really take very little time to do the rebuild, but that's not the case.
> 
> As we said last time Bjorn, it's the stage that looks like "CC [M[ drivers/foo/var/baz.mod.o" that takes forever. I am not sure what the stage is called though...

(First I thought it was missing the `-z none` but nope.) 

So the module build for clang is the step that takes forever? Are you getting ccache hits?

Edit! Discussed on IRC, and it's the linking that takes forever.

File: tests_rv/patch/build_rv64_clang_allmodconfig/build_rv64_clang_allmodconfig.sh [hunk 5]
```diff
@@ -1,23 +1,90 @@
...
+git checkout -q HEAD~
+
+echo "Building the tree before the patch"
+
+tuxmake --wrapper ccache --target-arch riscv -e PATH=$PATH --directory . \
+	--environment=KBUILD_BUILD_TIMESTAMP=@1621270510 \
+	--environment=KBUILD_BUILD_USER=tuxmake --environment=KBUILD_BUILD_HOST=tuxmake \
+	-o $tmpdir0 -b $tmpdir_b --toolchain llvm -z none --kconfig allmodconfig \
+	-K CONFIG_WERROR=n W=1 CROSS_COMPILE=riscv64-linux- \
+	config default \
```
Comment (bjoto) [inline tests_rv/patch/build_rv64_clang_allmodconfig/build_rv64_clang_allmodconfig.sh]: CONFIG_RANDSTRUCT_NONE=y missing -- this messes up the ccache.

File: tests_rv/patch/build_rv64_gcc_allmodconfig/build_rv64_gcc_allmodconfig.sh [hunk 6]
```diff
@@ -24,15 +25,16 @@ echo "Building the tree with the patch"
 tuxmake --wrapper ccache --target-arch riscv -e PATH=$PATH --directory . \
 	--environment=KBUILD_BUILD_TIMESTAMP=@1621270510 \
 	--environment=KBUILD_BUILD_USER=tuxmake --environment=KBUILD_BUILD_HOST=tuxmake \
-	-o $tmpdir0 --toolchain gcc -z none --kconfig allmodconfig -K CONFIG_WERROR=n \
-	-K CONFIG_GCC_PLUGINS=n W=1 \
+	-o $tmpdir_o -b $tmpdir_b --toolchain gcc -z none --kconfig allmodconfig \
+	-K CONFIG_WERROR=n \ -K CONFIG_GCC_PLUGINS=n W=1 CROSS_COMPILE=riscv64-linux- \
+	config default \
```
Comment (bjoto) [inline tests_rv/patch/build_rv64_gcc_allmodconfig/build_rv64_gcc_allmodconfig.sh:30]: Nit: remove quoted space before -K ...

**Review by bjoto — CHANGES_REQUESTED** [review]

**Comment by ConchuOD** [comment]
@bjoto So I had a bit of a think about the "leaked" builddir stuff and I came to the conclusion that leaking it is a bad idea.
I'm not sure that we can ever "trust" that someone's patchset does not do entirely botched things with the build system by hacking at kbuild. Having implemented a way to clean the build directories only at the start of a patch series I then ripped it out & put the rm $tmpdir_b back in!

Also, I disabled the clang build again temporarily. I want to build a PGOed toolchain as Nathan promises significant time savings from doing so. I'll probably kick that off tomorrow evening before I leave the office as his script will do a bunch of kernel builds to generate the profile & I'd rather that be the toolchain that we do clang/LLVM stuff with.

**Comment by bjoto** [comment]
Far enough! LGTM!

**Review by bjoto — APPROVED** [review]
LGTM!

**Pull request merged**